x/emissions/types/topic.go

```diff
@@ -0,0 +1,150 @@
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// Validate checks if the given CreateNewTopicRequest is valid
+func (msg *CreateNewTopicRequest) Validate(maxStringLen uint64) error {
+	if err := validateBech32(msg.Creator); err != nil {
+		return errors.Wrap(err, "invalid msg Creator address")
+	}
+
+	if len(msg.LossMethod) == 0 || uint64(len(msg.LossMethod)) > maxStringLen {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "loss method invalid")
+	}
+	if msg.EpochLength <= 0 {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "epoch length must be greater than zero")
+	}
+	if msg.WorkerSubmissionWindow == 0 {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "worker submission window must be greater than zero")
+	}
+	if msg.GroundTruthLag < msg.EpochLength {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "ground truth lag cannot be lower than epoch length")
+	}
+	if msg.WorkerSubmissionWindow > msg.EpochLength {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "worker submission window cannot be higher than epoch length")
+	}
+	if msg.AlphaRegret.Lte(alloraMath.ZeroDec()) || msg.AlphaRegret.Gt(alloraMath.OneDec()) || validateDec(msg.AlphaRegret) != nil {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "alpha regret must be greater than 0 and less than or equal to 1")
+	}
+	if msg.PNorm.Lt(alloraMath.MustNewDecFromString("2.5")) || msg.PNorm.Gt(alloraMath.MustNewDecFromString("4.5")) || validateDec(msg.PNorm) != nil {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "p-norm must be between 2.5 and 4.5")
+	}
+	if msg.Epsilon.Lte(alloraMath.ZeroDec()) || msg.Epsilon.IsNaN() || !msg.Epsilon.IsFinite() {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "epsilon must be greater than 0")
+	}
+	if uint64(len(msg.Metadata)) > maxStringLen {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "metadata invalid")
+	}
+	// no validation on AllowNegative because either it is true or false
+	// and both are valid values
+	//	AllowNegative            bool
+
+	if !isAlloraDecZeroOrLessThanOne(msg.MeritSortitionAlpha) {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "merit sortition alpha must be between 0 and 1 inclusive")
+	}
+	if !isAlloraDecZeroOrLessThanOne(msg.ActiveInfererQuantile) {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "active inferer quantile must be between 0 and 1 inclusive")
+	}
+	if !isAlloraDecZeroOrLessThanOne(msg.ActiveForecasterQuantile) {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "active forecaster quantile must be between 0 and 1 inclusive")
+	}
+	if !isAlloraDecZeroOrLessThanOne(msg.ActiveReputerQuantile) {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "active reputer quantile must be between 0 and 1 inclusive")
+	}
+
+	return nil
+}
+
+// Validate checks if the given Topic is valid
+func (topic Topic) Validate(params Params) error {
+	if topic.Id == 0 {
+		return errors.Wrap(sdkerrors.ErrInvalidType, "topic id zero is reserved")
```

Comment:
do we really want this amount of redundancy? Topics just can be set by using the createTopic which already has basically the same validation.

Reply:
Defense in depth 

I think the approach ought to be, do all this redundant stuff get the chain bulletproof, then later go back with an optimization pass

Reply:
Agree, we can always roll back later